### PR TITLE
[FLINK-11844][table-api] Simplify over window API classes and improve documentation

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR removes duplicate code and thus simplifies the over window API classes. It improves the documentation as well.

## Brief change log

- Duplicate classes removed
- Docs about windows updated

## Verifying this change

This change is already covered by existing tests, such as OverWindowITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by twalthr** [comment]
@sunjincheng121 Would be great if you can take a quick look.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala [hunk 1]
```diff
@@ -20,90 +20,180 @@ package org.apache.flink.table.api
...
     * Set the following offset (based on time or row-count intervals) for over window.
     *
-    * @param following following offset that relative to the current row.
-    * @return this over window
+    * @param followingExpr following offset that relative to the current row.
+    * @return an over window with defined following
     */
-  def following(following: String): OverWindowWithPreceding = {
-    this.following(ExpressionParser.parseExpression(following))
+  def following(followingExpr: String): OverWindowPartitionedOrderedPreceding = {
```
Comment (hequn8128) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: Should we rename these xxxExpr to xxx? It seems strange to add the "Expr" suffix for the String parameters.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/windows.scala [hunk 2]
```diff
@@ -85,73 +85,36 @@ object Session {
...
     *
-    * @param partitionBy some partition keys.
-    * @return A partitionedOver instance that only contains the orderBy method.
+    * @param orderBy field reference
+    * @return an over window with defined order
     */
-  def partitionBy(partitionBy: Expression*): PartitionedOver = {
-    PartitionedOver(partitionBy.toArray)
+  def orderBy(orderBy: Expression): OverWindowPartitionedOrdered = {
+    new OverWindowPartitionedOrdered(Seq[Expression](), orderBy)
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/windows.scala]: `Seq[Expression]()` -> `Seq()` may be works well. But feel free to keep current state.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala [hunk 3]
```diff
@@ -20,90 +20,180 @@ package org.apache.flink.table.api
 
 import org.apache.flink.table.expressions._
 import org.apache.flink.table.plan.logical._
-import org.apache.flink.table.typeutils.{RowIntervalTypeInfo, TimeIntervalTypeInfo}
-import org.apache.flink.table.api.scala.{CURRENT_RANGE, CURRENT_ROW}
 
 /**
-  * Over window is similar to the traditional OVER SQL.
+  * A over-window specification.
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: If we using `over-window` we should keep consistent throughout the documentation comments. And i think `over-window` is better than `over window` in the Over-window definition.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala [hunk 4]
```diff
@@ -20,90 +20,180 @@ package org.apache.flink.table.api
...
+    *
+    * For streaming tables, reference a rowtime or proctime time attribute here
+    * to specify the time mode.
+    *
+    * For batch tables, refer to a timestamp or long attribute.
+    *
+    * @param orderByExpr field reference
+    * @return an over window with defined order
+    */
+  def orderBy(orderByExpr: String): OverWindowPartitionedOrdered = {
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: orderByExpr -> orderBy

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala [hunk 5]
```diff
@@ -20,90 +20,180 @@ package org.apache.flink.table.api
...
-    private[flink] val following: Expression)
+class OverWindowPartitionedOrdered(partitionBy: Seq[Expression], orderBy: Expression) {
+
+  /**
+    * Set the preceding offset (based on time or row-count intervals) for over window.
+    *
+    * @param precedingExpr preceding offset relative to the current row.
+    * @return an over window with defined preceding
+    */
+  def preceding(precedingExpr: String): OverWindowPartitionedOrderedPreceding = {
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: precedingExpr -> preceding

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala (hunk 1 above)
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: +1

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala (hunk 3 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: Thanks for the hint. Yes, we should be consistent. Actually, this was the only remaining of `over-window`. I changed it to `over window`. Now we should be consistent with `tumbling windows, session windows` etc.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: +1, my IDE was just complaining about the name duplicate but I solved this issue by using `this.` again.

**Comment by twalthr** [comment]
Thanks for the review @hequn8128 and @sunjincheng121. I addressed your comments. Will merge this now...

**Comment by twalthr** [comment]
@walterddr good point with the documentation. I will change it to `optional` and add a sentence about the default behavior.

**Comment by twalthr** [comment]
@walterddr the documentation for Flink 1.7 is correct. This behavior has been changed recently for Flink 1.8 (see FLINK-10676).

**Pull request closed without merging**